File: src/forge/github.py
"""GitHub source — wraps the gh CLI to interact with GitHub issues."""

from __future__ import annotations

import json
import re
import subprocess
from dataclasses import dataclass


@dataclass
class GitHubIssue:
    """A single sub-issue from a GitHub PRD."""

    number: int
    title: str
    body: str


class GitHubSource:
    """Reads and manages issues from a GitHub PRD."""

    def __init__(self, prd_number: int) -> None:
        self.prd_number = prd_number
        self._validate_gh_cli()
        self.owner, self.repo = self._get_owner_repo()
        self._validate_prd()

    @staticmethod
    def _validate_gh_cli() -> None:
        """Check that gh CLI is installed and authenticated."""
        try:
            subprocess.run(["gh", "--version"], capture_output=True, check=True)
        except FileNotFoundError:
            raise SystemExit(
                "Error: gh CLI is not installed. See https://cli.github.com"
            )
        result = subprocess.run(
            ["gh", "auth", "status"], capture_output=True, text=True
        )
        if result.returncode != 0:
            raise SystemExit(
                "Error: gh CLI is not authenticated. Run 'gh auth login'."
            )

    @staticmethod
    def _get_owner_repo() -> tuple[str, str]:
        """Extract owner/repo from the git remote origin URL."""
        result = subprocess.run(
            ["git", "remote", "get-url", "origin"],
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            raise SystemExit("Error: could not determine git remote origin.")
        url = result.stdout.strip()

        # SSH: git@github.com:owner/repo.git
        match = re.match(r"git@github\.com:([^/]+)/([^/]+?)(?:\.git)?$", url)
        if match:
            return match.group(1), match.group(2)

        # HTTPS or SSH protocol: https://github.com/owner/repo.git
        match = re.match(
            r"(?:https?|ssh)://[^/]*github\.com/([^/]+)/([^/]+?)(?:\.git)?$", url
        )
        if match:
            return match.group(1), match.group(2)

        raise SystemExit(
            f"Error: could not parse owner/repo from remote URL: {url}"
        )

    def _validate_prd(self) -> None:
        """Validate that the PRD issue exists and has the 'prd' label."""
        result = subprocess.run(
            [
                "gh", "issue", "view", str(self.prd_number),
                "--json", "labels", "-q", ".labels[].name",
            ],
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            raise SystemExit(
                f"Error: GitHub issue #{self.prd_number} not found."
            )
        labels = [label for label in result.stdout.strip().split("\n") if label]
        if "prd" not in labels:
            raise SystemExit(
                f"Error: Issue #{self.prd_number} does not have the 'prd' label."
            )

    def get_prd_content(self) -> str:
        """Fetch and return the PRD issue body."""
        result = subprocess.run(
            [
                "gh", "issue", "view", str(self.prd_number),
                "--json", "body", "-q", ".body",
            ],
            capture_output=True,
            text=True,
            check=True,
        )
        return result.stdout.strip()

    def _fetch_sub_issues(self) -> list[dict]:
        """Fetch sub-issues of the PRD via the GraphQL API."""
        query = """
query($owner: String!, $repo: String!, $number: Int!) {
  repository(owner: $owner, name: $repo) {
    issue(number: $number) {
      subIssues(first: 100) {
        nodes {
          number
          title
          state
          body
        }
      }
    }
  }
}"""
        result = subprocess.run(
            [
                "gh", "api", "graphql",
                "-F", f"owner={self.owner}",
                "-F", f"repo={self.repo}",
                "-F", f"number={self.prd_number}",
                "-f", f"query={query}",
            ],
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            raise SystemExit(
                f"Error fetching sub-issues: {result.stderr.strip()}"
            )
        data = json.loads(result.stdout)
        return data["data"]["repository"]["issue"]["subIssues"]["nodes"]

    def get_next_issue(self) -> GitHubIssue | None:
        """Find the next open sub-issue sorted by number, or None if all complete."""
        sub_issues = self._fetch_sub_issues()
        open_issues = [i for i in sub_issues if i["state"] == "OPEN"]
        if not open_issues:
            return None
        open_issues.sort(key=lambda i: i["number"])
        first = open_issues[0]
        return GitHubIssue(
            number=first["number"],
            title=first["title"],
            body=first.get("body") or "",
        )

    def get_remaining_count(self) -> tuple[int, int]:
        """Return (remaining, total) sub-issue counts."""
        sub_issues = self._fetch_sub_issues()
        total = len(sub_issues)
        remaining = sum(1 for i in sub_issues if i["state"] == "OPEN")
        return remaining, total

    def complete_issue(self, issue: GitHubIssue) -> None:
        """Close the GitHub issue."""
        subprocess.run(
            ["gh", "issue", "close", str(issue.number)],
            capture_output=True,
            text=True,
            check=True,
        )

File: src/github.ts
/** GitHub source — wraps the gh CLI to interact with GitHub issues. */

import type { GitHubIssue, GitHubPrd, Source } from "./types";

function runSync(cmd: string[]): { exitCode: number; stdout: string; stderr: string } {
	const proc = Bun.spawnSync(cmd, {
		stdout: "pipe",
		stderr: "pipe",
	});
	return {
		exitCode: proc.exitCode,
		stdout: proc.stdout.toString(),
		stderr: proc.stderr.toString(),
	};
}

export class GitHubSource implements Source<GitHubIssue> {
	readonly prdNumber: number;
	readonly owner: string;
	readonly repo: string;

	constructor(prdNumber: number) {
		this.prdNumber = prdNumber;
		GitHubSource.validateGhCli();
		const [owner, repo] = GitHubSource.getOwnerRepo();
		this.owner = owner;
		this.repo = repo;
		this.validatePrd();
	}

	static validateGhCli(): void {
		try {
			runSync(["gh", "--version"]);
		} catch {
			throw new Error("Error: gh CLI is not installed. See https://cli.github.com");
		}
		const result = runSync(["gh", "auth", "status"]);
		if (result.exitCode !== 0) {
			throw new Error("Error: gh CLI is not authenticated. Run 'gh auth login'.");
		}
	}

	static getOwnerRepo(): [string, string] {
		const result = runSync(["git", "remote", "get-url", "origin"]);
		if (result.exitCode !== 0) {
			throw new Error("Error: could not determine git remote origin.");
		}
		const url = result.stdout.trim();

		// SSH: git@github.com:owner/repo.git
		const sshMatch = url.match(/^git@github\.com:([^/]+)\/([^/]+?)(?:\.git)?$/);
		if (sshMatch) {
			return [sshMatch[1], sshMatch[2]];
		}

		// HTTPS or SSH protocol: https://github.com/owner/repo.git
		const httpsMatch = url.match(
			/^(?:https?|ssh):\/\/[^/]*github\.com\/([^/]+)\/([^/]+?)(?:\.git)?$/,
		);
		if (httpsMatch) {
			return [httpsMatch[1], httpsMatch[2]];
		}

		throw new Error(`Error: could not parse owner/repo from remote URL: ${url}`);
	}

	validatePrd(): void {
		const result = runSync([
			"gh",
			"issue",
			"view",
			String(this.prdNumber),
			"--json",
			"labels",
			"-q",
			".labels[].name",
		]);
		if (result.exitCode !== 0) {
			throw new Error(`Error: GitHub issue #${this.prdNumber} not found.`);
		}
		const labels = result.stdout
			.trim()
			.split("\n")
			.filter((l) => l);
		if (!labels.includes("prd")) {
			throw new Error(`Error: Issue #${this.prdNumber} does not have the 'prd' label.`);
		}
	}

	async getPrdContent(): Promise<string> {
		const prd = this.getPrd();
		return prd.body;
	}

	getPrd(): GitHubPrd {
		const result = runSync(["gh", "issue", "view", String(this.prdNumber), "--json", "title,body"]);
		if (result.exitCode !== 0) {
			throw new Error(`Error: failed to fetch PRD issue #${this.prdNumber}.`);
		}
		const data = JSON.parse(result.stdout);
		return {
			number: this.prdNumber,
			title: (data.title ?? "").trim(),
			body: (data.body ?? "").trim(),
		};
	}

	private fetchSubIssues(): Array<{
		number: number;
		title: string;
		state: string;
		body: string;
	}> {
		const query = `
query($owner: String!, $repo: String!, $number: Int!) {
  repository(owner: $owner, name: $repo) {
    issue(number: $number) {
      subIssues(first: 100) {
        nodes {
          number
          title
          state
          body
        }
      }
    }
  }
}`;
		const result = runSync([
			"gh",
			"api",
			"graphql",
			"-F",
			`owner=${this.owner}`,
			"-F",
			`repo=${this.repo}`,
			"-F",
			`number=${this.prdNumber}`,
			"-f",
			`query=${query}`,
		]);
		if (result.exitCode !== 0) {
			throw new Error(`Error fetching sub-issues: ${result.stderr.trim()}`);
		}
		const data = JSON.parse(result.stdout);
		return data.data.repository.issue.subIssues.nodes;
	}

	async getNextIssue(): Promise<GitHubIssue | null> {
		const subIssues = this.fetchSubIssues();
		const openIssues = subIssues
			.filter((i) => i.state === "OPEN")
			.sort((a, b) => a.number - b.number);
		if (openIssues.length === 0) {
			return null;
		}
		const first = openIssues[0];
		return {
			number: first.number,
			title: first.title,
			body: first.body ?? "",
		};
	}

	async getRemainingCount(): Promise<[number, number]> {
		const subIssues = this.fetchSubIssues();
		const total = subIssues.length;
		const remaining = subIssues.filter((i) => i.state === "OPEN").length;
		return [remaining, total];
	}

	async completeIssue(issue: GitHubIssue): Promise<void> {
		const result = runSync(["gh", "issue", "close", String(issue.number)]);
		if (result.exitCode !== 0) {
			throw new Error(`Error: failed to close issue #${issue.number}: ${result.stderr.trim()}`);
		}
	}
}
